app/models/user.rb

```diff
@@ -56,21 +56,41 @@
 class User < ApplicationRecord
   devise :database_authenticatable, :lockable, :validatable, :timeoutable, :trackable, :invitable, :recoverable
 
-  belongs_to :vita_partner, optional: true
   has_many :assigned_tax_returns, class_name: "TaxReturn", foreign_key: :assigned_user_id
+
+  #
+  belongs_to :vita_partner, optional: true
   has_and_belongs_to_many :supported_organizations,
            join_table: "users_vita_partners",
            class_name: "VitaPartner"
+  #
+
+  has_many :memberships
+  accepts_nested_attributes_for :memberships
 
   attr_encrypted :access_token, key: ->(_) { EnvironmentCredentials.dig(:db_encryption_key) }
 
   validates_presence_of :name
   validates_inclusion_of :timezone, in: ActiveSupport::TimeZone.country_zones("us").map { |tz| tz.tzinfo.name }
 
+  # returns true if the user has role "lead" for organization or its parents.
+  # @param keyword [Boolean] exclude_ancestors
+  def can_lead?(vita_partner, exclude_ancestors: false)
+    memberships.where({
+        role: "lead",
+        vita_partner_id: [
+          vita_partner.id,
+          *(vita_partner.ancestors unless exclude_ancestors)
+        ].compact
+    }).exists?
+  end
+
+  # returns organizations a user is a member of, and their children (if any).
   def accessible_organizations
-    accessible_organization_ids = [vita_partner_id] + supported_organizations.pluck(:id)
-    VitaPartner.where(id: accessible_organization_ids).or(
-      VitaPartner.where(parent_organization_id: accessible_organization_ids)
+    member_org_ids = memberships.pluck(:vita_partner_id)
+    VitaPartner.where(id: member_org_ids).or(
+      VitaPartner.where(parent_organization_id: member_org_ids)
     )
   end
```

Comment:
This method is complicated by keeping vita_partner and supported_organizations in the model. 

Additionally`pluck` is a method on a query that returns an array.
This would be more performant for any downstream actions if it returned an active record query of VitaPartner.

```ruby
# up above
has_many :vita_partners, through: :memberships
# without supported_organizations or vita_partner, and with
def accessible_organizations
  vita_partners.or(VitaPartner.where(parent_organization: vita_partners)
end
```

Reply:
One caveat: supported_organizations and vita_partner are not part of this query. We removed that and we're only getting the vita_partner_ids from the memberships. Supported organizations and vita_partner should no longer be referenced directly in the tests, abilities, or the majority of views we found (a more thorough search would happen when we remove those associations completely in a followup PR). The optimization is definitely a good idea -- Alex and I talked about adding a has_many :vita_partners, through: :memberships, but because of the parent/child structure of grabbing the organizations, we didn't want to confuse people with the meaning of "vita_partners" vs "accessible_organizations". However, I'm willing to add it because it is a ncie optimization.